auth/src/services/aws/dynamodb.ts

```diff
@@ -0,0 +1,553 @@
+import { didRegex } from '../../utils/did.js'
+import { now } from '../../utils/datetime.js'
+
+import {
+    ConfigKey,
+    createEmailVerificationCode,
+    Database,
+    DIDResult,
+    DIDStatus,
+    OTPStatus
+} from '../db.js'
+import { KeyService } from "../key.js"
+
+const AWS_REGION = process.env.AWS_REGION ?? ''
+const INITIAL_NONCE = '0'
+
+/**
+ * Returns current timestamp as unix integer
+ * @returns Unix timestamp
+ */
+const CONFIG_TABLE_NAME = process.env.IS_OFFLINE ? `cas-auth-config-${now()}` : (process.env.DB_CONFIG_TABLE_NAME ?? '')
+const DID_TABLE_NAME = process.env.IS_OFFLINE ? `cas-auth-did-${now()}` : (process.env.DB_DID_TABLE_NAME ?? '')
+const OTP_TABLE_NAME = process.env.IS_OFFLINE ? `cas-auth-otp-${now()}` : (process.env.DB_OTP_TABLE_NAME ?? '')
+
+class ItemNotFoundError extends Error { }
+
+export type Item = Record<string, AttributeValue>
+type ExpressionAttributeValues = Item
+
+export class DynamoDB implements Database {
+    name: string
+    readonly client: DynamoDBClient
+    readonly keyService: KeyService
+    private readonly _shouldCreateTableIfNotExists: boolean
+
+    constructor(createTableIfNotExists: boolean) {
+        this._shouldCreateTableIfNotExists = createTableIfNotExists
+        if (CONFIG_TABLE_NAME == '') {
+            throw Error('Missing DB_CONFIG_TABLE_NAME')
+        }
+        if (DID_TABLE_NAME == '') {
+            throw Error('Missing DB_DID_TABLE_NAME')
+        }
+        if (OTP_TABLE_NAME == '') {
+            throw Error('Missing DB_OTP_TABLE_NAME')
+        }
+        this.client = process.env.IS_OFFLINE
+            ? new DynamoDBClient({ endpoint: 'http://localhost:8000' })
+            : new DynamoDBClient({ region: AWS_REGION })
+    }
+
+    async init() {
+        for (const tableName of [CONFIG_TABLE_NAME, DID_TABLE_NAME, OTP_TABLE_NAME]) {
+            if (this._shouldCreateTableIfNotExists) {
+                await this._createTableIfNotExists(tableName)
+            } else {
+                const shouldThrow = true
+                await this._checkTableExists(tableName, shouldThrow)
+            }
+        }
+    }
+
+    private async _createTableIfNotExists(tableName: string): Promise<void> {
+        const shouldthrow = process.env.IS_OFFLINE ? false : true
+        if (await this._checkTableExists(tableName, shouldthrow)) return
+
+        const input: CreateTableCommandInput = {
+            TableName: tableName,
+            KeySchema: [
+                { AttributeName: 'PK', KeyType: 'HASH' },
+                { AttributeName: 'SK', KeyType: 'RANGE' }
+            ],
+            AttributeDefinitions: [
+                { AttributeName: 'PK', AttributeType: 'S' },
+                { AttributeName: 'SK', AttributeType: 'S' },
+                { AttributeName: 'GSI-1-PK', AttributeType: 'S' },
+                { AttributeName: 'GSI-1-SK', AttributeType: 'S' },
+            ],
+            BillingMode: BillingMode.PAY_PER_REQUEST,
```

Comment:
Is this the optimal billing mode?

Reply:
It seems straight forward to start. I think we could do provisioned but not sure how much would be appropriate. Suggestions are welcome!